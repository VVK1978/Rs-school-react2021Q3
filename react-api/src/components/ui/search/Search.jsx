import React from 'react';
import getSearchCharacters from '../../../services/api.jsx';
import Button from '../button/button.jsx';
import Input from '../input/Input.jsx';

export default function Search(props) {
  const { data } = props;
  const { updateState } = props;
  const handleSubmit = async (event) => {
    event.preventDefault();
    // eslint-disable-next-line no-unused-expressions
    data.isFetching
      ? (document.querySelector('.loader').style.display = 'block')
      : '';
    const searchValue = event.target.search.value;
    data.searchValue = searchValue;
    const response = await getSearchCharacters(data);
    const totalPages = await response.pages;
    const characters = await response.docs;
    const updatedData = {
      ...data,
      searchValue,
      characters,
      totalPages,
    };
    updateState(updatedData);
  };

  return (
    <form className="search-form" onSubmit={handleSubmit} name="search-form">
      <Input />
      <Button />
    </form>
  );
}
