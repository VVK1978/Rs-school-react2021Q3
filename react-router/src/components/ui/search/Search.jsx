import React from 'react';
import Button from '../button/button.jsx';
import Input from '../input/Input.jsx';

export default function Search() {
  const handleSubmit = (event) => {
    event.preventDefault();
  };

  return (
    <>
      <form action="#" className="search-form" onSubmit={handleSubmit}>
        <Input />
        <Button />
      </form>
    </>
  );
}
