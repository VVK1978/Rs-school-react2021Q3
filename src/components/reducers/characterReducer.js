const SET_CHARACTERS = 'SET_CHARACTERS';
const SET_IS_FETCHING = 'SET_IS_FETCHING';
const SET_PER_PAGE = 'SET_PER_PAGE';
const SET_CURRENT_PAGE = 'SET_CURRENT_PAGE';
const SET_SEARCH_VALUE = 'SET_SEARCH_VALUE';
const SET_SORT_VALUE = 'SET_SORT_VALUE';
const SET_SORT_DIRECTION = 'SET_SORT_DIRECTION';

const defaultState = {
  items: [],
  isFetching: true,
  currentPage: 1,
  perPage: 10,
  totalPages: 0,
  searchValue: '',
  sortBy: 'name',
  sortDirection: 'asc',
};
export default function characterReducer(state = defaultState, action) {
  switch (action.type) {
    case SET_CHARACTERS:
      return {
        ...state,
        items: action.payload.docs,
        totalPages: action.payload.pages,
        isFetching: false,
      };
    case SET_IS_FETCHING:
      return {
        ...state,
        isFetching: action.payload,
      };
    case SET_CURRENT_PAGE:
      return {
        ...state,
        currentPage: action.payload,
      };
    case SET_PER_PAGE:
      return {
        ...state,
        perPage: action.payload,
      };
    case SET_SEARCH_VALUE:
      return {
        ...state,
        searchValue: action.payload,
      };
    case SET_SORT_VALUE:
      return {
        ...state,
        sortBy: action.payload,
      };
    case SET_SORT_DIRECTION:
      return {
        ...state,
        sortDirection: action.payload,
      };
    default:
      return state;
  }
}

export const setCharacters = (characters) => ({
  type: SET_CHARACTERS,
  payload: characters,
});

export const setIsFetching = (bool) => ({
  type: SET_IS_FETCHING,
  payload: bool,
});

export const setCurrentPage = (page) => ({
  type: SET_CURRENT_PAGE,
  payload: page,
});

export const setPerPage = (perPage) => ({
  type: SET_PER_PAGE,
  payload: perPage,
});

export const setSearchValue = (searchValue) => ({
  type: SET_SEARCH_VALUE,
  payload: searchValue,
});

export const setSort = (sortBy) => ({
  type: SET_SORT_VALUE,
  payload: sortBy,
});

export const setSortDirection = (sortDirection) => ({
  type: SET_SORT_DIRECTION,
  payload: sortDirection,
});
