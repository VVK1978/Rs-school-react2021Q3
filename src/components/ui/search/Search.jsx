/* eslint-disable no-undef */
import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getSearchCharacters } from '../../../actions/characters';
import Button from '../button/button';
import Input from '../input/Input';

export default function Search() {
  const searchValue = useSelector((state) => state.characters.searchValue);
  const perPage = useSelector((state) => state.characters.perPage);
  const sortBy = useSelector((state) => state.characters.sortBy);
  const sortDirection = useSelector((state) => state.characters.sortDirection);

  const dispatch = useDispatch();
  const handleSubmit = (event) => {
    event.preventDefault();
    if (document.getElementById('loader')) {
      document.getElementById('loader').classList.add('active');
    }
    dispatch(
      getSearchCharacters(searchValue, perPage, 1, sortBy, sortDirection),
    );
  };

  return (
    <>
      <form className="search-form" onSubmit={handleSubmit} name="search-form">
        <Input />
        <Button />
      </form>
    </>
  );
}
