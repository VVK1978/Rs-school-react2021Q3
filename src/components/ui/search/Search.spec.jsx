/* eslint-disable no-undef */
import shallow from 'enzyme/build/shallow';
import React from 'react';
import * as redux from 'react-redux';
import { createStore } from 'redux';
import Search from './Search.jsx';
import characterReducer from '../../reducers/characterReducer';

const store = createStore(characterReducer);
describe('Component with redux hooks', () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let mockDispatch;

  beforeEach(() => {
    spyOnUseSelector = jest.spyOn(redux, 'useSelector');
    spyOnUseSelector.mockReturnValue([{ id: 1, text: 'Old Item' }]);
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render Search component', () => {
    const component = shallow(<Search />);
    expect(component).toMatchSnapshot();
  });

  it('should Search value state', () => {
    const { searchValue } = store.getState();
    expect(store.getState().searchValue).toEqual(searchValue);
  });

  it('should Search per page state', () => {
    const { perPage } = store.getState();
    expect(store.getState().perPage).toEqual(perPage);
  });
});
