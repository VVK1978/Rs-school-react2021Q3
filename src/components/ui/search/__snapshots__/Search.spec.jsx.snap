// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component with redux hooks should render Search component 1`] = `ShallowWrapper {}`;
